lib/ocamlorg_web/graphql.ml

```diff
@@ -6,59 +6,125 @@ type package_info =
+  match res with
+  | false ->
+    "wrong_limit"
+  | _ ->
+    let res = is_valid_offset in
+    (match res with false -> "wrong_offset" | _ -> "true")
+
+let packages_list contains offset limit all_packages t =
+  match contains with
+  | None ->
+    List.filteri (fun i _ -> offset <= i && i < offset + limit) all_packages
+  | Some letters ->
+    List.filteri
+      (fun i _ -> offset <= i && i < offset + limit)
+      (Package.search_package t letters)
+
+let all_packages_result contains offset limit all_packages t =
+  let total_packages = List.length all_packages in
+  let limit = match limit with None -> total_packages | Some limit -> limit in
+  let result = is_valid_params limit offset total_packages in
+  match result with
+  | "wrong_offset" ->
+    Error
+      ("offset must be greater than or equal to 0 AND less than or equal to "
+      ^ string_of_int total_packages)
+  | "wrong_limit" ->
+    Error
+      ("limit must be greater than or equal to 0 AND less than or equal to "
+      ^ string_of_int total_packages)
+  | _ ->
+    let packages = packages_list contains offset limit all_packages t in
+    Ok { total_packages; packages }
+
+let package_result name version t =
+  match version with
+  | None ->
+    let package = Package.get_package_latest t (Package.Name.of_string name) in
+    (match package with
+    | None ->
+      Error ("No package matching " ^ name ^ " was found")
+    | Some package ->
+      Ok package)
+  | Some version ->
+    let package =
+      Package.get_package
+        t
+        (Package.Name.of_string name)
+        (Package.Version.of_string version)
+    in
+    (match package with
+    | None ->
+      Error ("No package matching " ^ name ^ ": " ^ version ^ " was found")
+    | Some package ->
+      Ok package)
+
+let package_versions_result name from upto t =
+  let all_packages = Package.all_packages_latest t in
+  let total_packages = List.length all_packages in
+  let packages =
+    Package.get_packages_with_name t (Package.Name.of_string name)
+  in
+  match packages with
+  | None ->
+    Error ("No package matching " ^ name ^ " was found")
+  | Some packages ->
+    let total_package_versions = List.length packages in
+    if from = None && upto = None then
+      Ok { total_packages; packages }
+    else
+      let from =
+        match from with
+        | None ->
+          Package.Version.to_string (Package.version (List.hd packages))
+        | Some from ->
+          from
+      in
+      let upto =
+        match upto with
+        | None ->
+          Package.Version.to_string
+            (Package.version (List.nth packages (total_package_versions - 1)))
+        | Some upto ->
+          upto
+      in
+      let package_list =
+        List.filter
+          (fun package ->
+            is_in_range
+              (Package.version package)
+              (Package.Version.of_string from)
+              (Package.Version.of_string upto))
+          packages
```

Comment:
From your emoji reaction I assume that you don't like that suggestion very much? I think the main point of PR reviews is to discuss the code and to discuss different ideas. So it's cool if you have a different opinion and prefer to keep it the way it is now. It would be very helpful if you could point out why you prefer to keep it the way it is now.

Reply:
Not an angry emoji, it's a confused emoji. I meant that I did not understand it very well. I apologize for the misunderstanding

Reply:
Ah ok, sorry, I misunderstood the emoji. What I mean is the following: currently, `is_in_range` is defined as
```
let is_in_range current_version from_version upto_version =
  current_version >= from_version && current_version <= upto_version
```
Instead you could define it as something like
```
let is_in_range current_version from_version upto_version =
  match from_version, upto_version with
  | None, None -> true
  | Some from_version, None ->  current_version >= from_version
  | None, Some upto_version -> current_version <= upto_version
  | Some from_version, Some uptp_version -> current_version >= from_version && current_version <= upto_version
```
Then you could directly pass `from` and `upto` as options to that function without unwrapping them:
```
            is_in_range
              (Package.version package)
              (Option.map Package.Version.of_string from)
              (Option.map Package.Version.of_string upto))
```
Just as a proposal.

Reply:
Nice! I think it's quite a bit easier to read now, thanks!

Reply:
Just one last comment about this: if you wanted to, in your tests you could now also test the `is_in_range` function passing in `None` for `from` and/or `upto`. But it's not necessary. It's up to you if you want to do that :)